### PR TITLE
CDP replacement candidate: direct python API for BiDi

### PR DESCRIPTION
**Pull request by martinpitt**
This is another candidate for https://issues.redhat.com/browse/COCKPIT-1139 , in the similar spirit as #926

This talks to https://w3c.github.io/webdriver-bidi/ directly, requires *zero* new npm dependencies (in fact, we can drop chrome-remote-interface), and only a new package (`chromedriver`) in the tasks container (~installed locally for this demo~ added to official container now).

This is much further ahead than #926, in the sense of that it actually is a Python test, in our usual testlib shape. We won't need any extra `node` process to talk to a JS webdriver module any more, and they aren't any good ones for BiDi anyway.

Compared to CDP this has the following advantages:
 * Fewer moving parts and less code
 * At the API level, Firefox and Chromium behave the same, so we can drop our `{firefox,chromium}-cdp-driver.js` forks
 * Very low input API level, we can even simulate touch swipes and others. With CDP we just synthesize a `MouseEvent`, while this *actually* goes through the browser.

Hiccups:
 - [x] Land and deploy https://github.com/cockpit-project/cockpituous/pull/622 to use the packaged chromedriver (#935 ), then drop the local download
 - [x]  that `input.performActions` does not understand key names like "Return" or "ArrowUp".  https://github.com/w3c/webdriver-bidi/issues/757
 - [x] clicks in frames miss their target in chrome (works fine with Firefox)
 - [x] When showing the browser, the test sometimes crashes with something like "target is outside of viewport". needs scrollIntoView() or so (I remember this from Selenium)
 - [x] Check how to enable coverage (if no API, then via a single CDP call or some other hack)
 - [x] Test loading/using sizzle

https://issues.redhat.com/browse/COCKPIT-1154

### CODE REVIEW

**Comment by martinpitt** [comment]
@jelly @allisonkarlitskaya At this point it seems pretty clear to me that this approach is much better than #926 (where all of the required JS → Python glue isn't even written, it uses the outdated webdriver protocol, and requires umpteen new dependencies, and it's also slower).

This is demo code, so no need for a minutely review, but I'd like to hear your coarse-grained opinion about the approach. Thanks!

File: bidi.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+@dataclass
+class LogMessage:
+    level: str  # like "info"
+    type: str  # usually "console"
+    timestamp: int
+    args: list[object]
+    text: str
+
+    def __init__(self, message_params):
```
Comment (allisonkarlitskaya) [inline bidi.py:42]: How do you feel about the cockpit json helpers?  I notice you have your own `JsonObject` defined above, and indeed, we can't import `jsonutil` from `starter-kit`.  I wonder if we should move those helpers to a more neutral location where they can also get used by tests.

It's a bit of a weird situation.  We might end up with two copies... or maybe we don't have to care about typing too much here.

File: bidi.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        self.task_reader.cancel()
+        del self.task_reader
+        await self.ws.close()
+        await self.close_bidi_session()
+        self.bidi_session.process.terminate()
+        await self.bidi_session.process.wait()
+        self.bidi_session = None
+        await self.http_session.close()
+
+    def ws_done_callback(self, future):
```
Comment (allisonkarlitskaya) [inline bidi.py:112]: It probably also makes sense to reject all outstanding commands here.  If something crashed, you don't want to get stuck awaiting a response that will never come.

File: bidi-test.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        else:
+            raise ValueError(f"unknown browser {browser}")
+        self.loop = asyncio.new_event_loop()
+        self.bidi_thread = threading.Thread(target=self.asyncio_loop_thread, args=(self.loop,), daemon=True)
+        self.bidi_thread.start()
+
+        asyncio.run_coroutine_threadsafe(self.driver.start_session(), self.loop).result()
+
+    def close(self):
+        asyncio.run_coroutine_threadsafe(self.driver.close(), self.loop).result()
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply bidi-test.py:46]: Thanks! I'm surprised that it surprised you :wink:

File: bidi.py (hunk 1 above)
Comment (martinpitt) [reply bidi.py:42]: Heh yes, had the same "wanna but can't" feeling. But TBH we just need it at exactly one place, I feel it's probably not worth sharing bridge with testlib code for that. Maybe if we run into more of these, but I don't see that coming -- the protocol side is basically "done", the remaining work is all on the "outwards" testlib side.

File: bidi.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+                    else:
+                        self.pending_commands[data["id"]].set_exception(
+                            WebdriverError(f"{data['type']}: {data['message']}"))
+                    del self.pending_commands[data["id"]]
+                    continue
+
+                if data["type"] == "event":
+                    if data["method"] == "log.entryAdded":
+                        log = LogMessage(data["params"])
+                        self.logs.append(log)
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply bidi.py:175]: Ah, sure -- cf. my question above with "does it already make sense at this point to use that exact API for the unit tests"? Do you have a pointer for what a unit test is trying to do that relies on waiting for log messages? There could certainly be one method for "set up a log watch future", and resolve it here if it's set up; is that enough?

File: bidi.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply bidi.py:112]: Good idea, thanks!

Comment (martinpitt) [reply bidi.py:112]: Help, I'm seeing double (bad github day?)

**Comment by martinpitt** [comment]
> I think the next thing I'd like to see is a version of this integrated into test/common/ in the cockpit repo....

Right, that's the plan. This is strictly "draft", mostly because it's a convenient place for a demo, and we wanted to compare different approaches.

File: bidi.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply bidi.py:112]: done.

File: bidi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply bidi.py:177]: See ./test/common/tap-cdp and `cdp.read_log()`

[the remark this answers is not in the document]
Comment (martinpitt) [reply bidi.py:177]: https://github.com/allisonkarlitskaya/cockpit/blob/test-server.py/test/pytest/webdriver_bidi.py#L81

**Comment by martinpitt** [comment]
Wow, it's very straightforward to talk to CDP. chromedriver already enables CDP and gives you the address in the returned capabilities. And looks like not only the bidi sessions/tabs appear there, they even have the *exact same* session IDs as bidi/webdriver:
```
❱❱❱ curl http://localhost:37773/json/list
[ {
   "description": "",
   "devtoolsFrontendUrl": "/devtools/inspector.html?ws=localhost:37773/devtools/page/9E5C285BA99CC696EE4FD2390D303029",
   "id": "9E5C285BA99CC696EE4FD2390D303029",
   "title": "fedora-40-127-0-0-2-2201",
   "type": "page",
   "url": "http://127.0.0.2:9091/",
   "webSocketDebuggerUrl": "ws://localhost:37773/devtools/page/9E5C285BA99CC696EE4FD2390D303029"
}, {
   "description": "",
   "devtoolsFrontendUrl": "/devtools/inspector.html?ws=localhost:37773/devtools/page/401D4B511DB168958EFF68E8F2F32087",
   "id": "401D4B511DB168958EFF68E8F2F32087",
   "title": "BiDi-CDP Mapper",
   "type": "page",
   "url": "data:,",
   "webSocketDebuggerUrl": "ws://localhost:37773/devtools/page/401D4B511DB168958EFF68E8F2F32087"
} ]
```

Unfortunately, `websocat ws://localhost:37773/devtools/page/9E5C285BA99CC696EE4FD2390D303029` immediately fails with "websocat: WebSocketError: I/O failure" in both default text and binary (`-b`) mode. However, it works fine with aiohttp, and I can send commands and receive replies. I just wonder about this:
```
DEBUG:bidi.proto:CDP ← '{"id": 0, "method": "Profiler.enable", "params": {}}'
DEBUG:bidi.proto:CDP → {'id': 0, 'result': {}}
DEBUG:bidi.proto:CDP ← '{"id": 1, "method": "Profiler.startPreciseCoverage", "params": {"callCount": false, "detailed": true}}'
DEBUG:bidi.proto:CDP → {'id': 1, 'error': {'code': -32000, 'message': 'Profiler is not enabled'}}
```
That's pretty much exactly what `CDP.start()` does in our current code.. :thinking: 

I am currently trying to only connect to the CDP websocket when necessary, it seems a big waste to let it run all the time -- then we have to read all the messages and discard them *sigh*. But that's my next attempt,

**Comment by martinpitt** [comment]
Look ma, CDP Profiler (aka coverage) support! That was my final TODO item that I could think of. Now I think it's time to aim higher and put it into actual cockpit.

**Comment by martinpitt** [comment]
This was just a  demo and has fulfilled its purpose. https://github.com/cockpit-project/cockpit/pull/20832 is Ze Real Zing.

**Pull request closed without merging**